examples/hif2a/fit_to_multiple_rbfes.py

```diff
@@ -277,93 +299,141 @@ def unflatten(theta: array) -> Dict[Handler, array]:
+            theta_increment = truncated_step(flat_theta, loss, flat_loss_grad,
+                                             step_lower_bound=_compute_step_lower_bound(loss, blown_up))
+            param_increments = unflatten_theta(theta_increment)
+
+            # for any parameter handler types being updated, update in place
+            for handle in ordered_handles:
+                handle_type = type(handle)
+                if handle_type in param_increments:
+                    print(f'updating {handle_type.__name__}')
+
+                    # TODO: careful -- this must be a "+=" or "-=" not an "="!
+                    handle.params += param_increments[handle_type]
+
+                    increment = param_increments[handle_type]
+                    increment = increment[increment != 0]
+                    min_update = 0.0
+                    max_update = 0.0
+                    if len(increment):
+                        min_update = np.min(increment)
+                        max_update = np.max(increment)
+                    # TODO: replace with a function that knows what to report about each handle type
+                    print(
+                        f'updated {len(increment)} params by between {min_update:.4f} and {max_update:.4f}')
+
+            t1 = time()
+            elapsed = t1 - t0
+
+            print(f'completed forcefield-updating step {step} in {elapsed:.3f} s !')
+
+            # save du_dls snapshot
+            path_to_du_dls = output_path.joinpath(f'du_dls_snapshot_{step}.npz')
+            print(f'saving du_dl trajs to {path_to_du_dls}')
+            du_dls_dict = dict()  # keywords here must be strings
+            for stage, results in results_this_step.items():
+                du_dls_dict[stage] = _results_to_arrays(results)[1]
+            np.savez(path_to_du_dls, **du_dls_dict)
+
+            # also save information about this step's parameter gradient and parameter update
+            # results to npz
+            path_to_npz = output_path.joinpath(f'theta_grad_loss_snapshot_{step}.npz')
+            print(f'saving theta, grad, loss snapshot to {path_to_npz}')
+            np.savez(
+                path_to_npz,
+                theta=np.array(flat_theta),
+                grad=np.array(flat_loss_grad),
+                loss=loss
+            )
+
+            # TODO: same for xs and du_dps snapshots
+
+            # save updated forcefield .py files after every gradient step
+            step_params = serialize_handlers(ff_handlers)
+            # TODO: consider if there's a more modular way to keep track of ff updates
+            _save_forcefield(output_path.joinpath(f"forcefield_checkpoint_{step}.py"), ff_params=step_params)
+            step += 1
+            if step >= args.param_updates:
+                break
+        if len(validation) > 0:
+            val_loss = np.zeros(len(validation))
```

Comment:
Saving out significantly less for the validation, wasn't clear to me if we want all of the save information or not.

Reply:
I think what you've done looks good. (No need to save params or gradient here, but it is helpful to save du_dls here, as you've implemented.) I think this has validation-set predictions computed after every epoch (rather than after every optimizer step), which I think makes sense as well.